### ISSUE TITLE
Example and explanation

### ISSUE THREAD

**Issue by platy** [Tamschi/lignin#7]
Hi! I'm interested in this project and especially how you've got a html renderer as well as a DOM differ. You explained that you built this because dodrio didn't fulfill your requirements, could you explain a bit more about how lignin differs from dodrio?

Also, I didn't find any example, I tried to use a dodrio example but there were a lot of differences.

Thanks!

**Comment by Tamschi** [comment]
Sure! Admittedly this crate is very badly documented right now, and I _may_ have to change the event handler signature a bit when I fix soundness for that in [Asteracea](https://github.com/Tamschi/Asteracea). Hopefully not, but just a heads-up.

- - -

The main difference, I think, is that `lignin` has meta nodes but isn't geared towards diffing being done in JS or directly against the DOM (so it's probably a bit slower than dodrio due to more web API calls from Rust!).

"Meta nodes" means that each node in the graph doesn't necessarily map to a DOM node, though the DOM node types are more directly represented than in dodrio: [lignin::Node#variants](https://docs.rs/lignin/0.0.5/lignin/enum.Node.html#variants).
The initial reason I made my own crate was because I couldn't get dodrio to render HTML comments, but the transparent `Multi` nodes mean diffing can be more efficient with dynamic collections or `if`s in the template. (I just render an empty multi node if the branch isn't taken.)
**Please ignore anything called "Remnant" for now.** Something I'd like to eventually support are DOM elements that linger after their VDOM representation is removed (which is great to allow for example for fade-out animations, without worrying too much about delayed state changes). This isn't implemented at all yet and I'll most likely have to redo some of the API there.

`lignin-html` [should be relatively self-explanatory](https://docs.rs/lignin-html/0.0.5/lignin_html/fn.render.html), though I think I could remove the bump allocator from the signature there 🤔 (It's a holdover from when I started this project, I think, which I forgot to remove while refactoring.)

- - -

The intended use for `lignin-dom` is with a double-buffered VDOM, so you keep the previous version around and then diff against that. The first version you need to read in from the browser DOM, since you (ideally) hydrate a static page in place.

I'll have to refer to some relatively old code of mine, and this definitely isn't ergonomic or ideal, but here's how you can run and update a web app with `lignin`:

```rust
let mut bump_a = Box::leak(Box::new(Bump::new()));
let mut bump_b = Box::leak(Box::new(Bump::new()));
let mut previous: Option<vNode> = None;
let container = // A DOM element reference
let render = Mutex::new(move || {
	let vdom_old = previous.take().unwrap_or_else(|| {
		vNode::Multi(bump_a.alloc_with(|| load_element(&container, &bump_a).content))
	});
	//TODO: Actually diff here.
	unsafe {
		//SAFETY:
		// This formally detaches the lifetime of `vdom` from `bump_b`, so that the latter can be swapped with `bump_a`.
		// Since the reset of `bump_a` only happens after `previous` is overwritten (and after the previous DOM has been updated away),
		// there shouldn't be any dangling references.
		update_child_nodes(&[vdom_old], &[], &container);
		let vdom: &Node = // Render VDOM here using `&bump_b`.
		update_child_nodes(&[], &[vdom], &container);
		previous = Some(transmute::<vNode<'_>, vNode<'_>>(vdom));
		bump_a.reset();
	}
	swap(&mut bump_a, &mut bump_b);
});
let render: &dyn Fn() = Box::leak(Box::new(move || {
	// Inefficient if called directly, should schedule a render instead.
	let render = &mut *render.lock().unwrap();
	render()
}));
```

A fixed version would look more like this right now:

```rust
let mut bump_a = Box::leak(Box::new(Bump::new()));
let mut bump_b = Box::leak(Box::new(Bump::new()));
let mut previous: Option<vNode> = None;
let container = // A DOM element reference
let render = Mutex::new(move || {
	let vdom_old = previous.take().unwrap_or_else(|| {
		vNode::Multi(bump_a.alloc_with(|| load_element(&container, &bump_a).content))
	});
	unsafe {
		//SAFETY:
		// This formally detaches the lifetime of `vdom` from `bump_b`, so that the latter can be swapped with `bump_a`.
		// Since the reset of `bump_a` only happens after `previous` is overwritten (and after the previous DOM has been updated away),
		// there shouldn't be any dangling references.
		let vdom: &Node = // Render VDOM here using `&bump_b`.
		update_child_nodes(&[vdom_old], &[vdom], &container); // <--
		previous = Some(transmute::<vNode<'_>, vNode<'_>>(vdom));
		bump_a.reset();
	}
	swap(&mut bump_a, &mut bump_b);
});
let render: &dyn Fn() = Box::leak(Box::new(move || {
	// Inefficient if called directly, should schedule a render instead.
	let render = &mut *render.lock().unwrap();
	render()
}));
```

…but `lignin-dom` is incomplete and likely to panic when doing this.

- - -

I've mainly been busy with getting [the template language](https://schichler.dev/Asteracea/) to a minimal useful state, so I haven't gotten around to making the entry point ergonomic yet. Sorry.

If you run into more specific issues, please tell me about it and I'll look into fixing them. A lot of this code is still from around when I was starting out with Rust, so I should be able to improve it pretty efficiently now.

I'll have to budget a bit regarding how much time I spend on it right now (since the crates are unstable, and stabilisation depends on hammering out various other crates I'm cooking in parallel), but if anything you'd like to use is unclear, I'll also look into adding documentation in that regard.

Also if you need a feature I haven't thought about, then please ask! I'm mainly developing this to make frontend apps with Asteracea, but the reason I spun out this package in the first place is that I'd like it to be helpful for other projects too.
No guarantees regarding actual implementation on my part, but at the very least I'll give you pointers about where to attach any necessary changes to the existing project.

**Comment by platy** [comment]
Thanks for the thorough reply!

I appreciate that you've published this early, I will definitely try it out

I'm interested in lignin as a base for interoperable app components, the lightweight design which doesn't assume anything about where it is being run seems ideal, and I especially like that you've already built the ability to parse the vdom from the dom. I feel like with rust I'm going to manage what I always hoped I would be able to do with javascript frontend / backend interop.

It's also interesting that you've thought about handling fade-out with remnants, I was working on an app doing animation last year, I refactored the whole drawing part to handle faded animation and then realized that fade-out wouldn't work without something like remnants.

The multi-nodes make a lot of sense too.

**Comment by Tamschi** [comment]
I'm glad that you like the concept, and to hear someone familiar with this topic thinks my ideas are sensible!

To be honest, my actual job experience in web development is (still) quite limited. I just happened to get into it as consultant on multiple (mostly very short) projects and ended up seeing a lot of pain points across the different frameworks/ecosystems 😓

I think our perspectives on the interop layer may be similar. Shared DTOs (and, admittedly, just not wanting to deal with wrong `*.d.ts` files at home) are the main reason I started looking into Rust for front-end web development.

It's just a proof of concept at this stage, but the (unpublished) CRUD app that escalated into my framework project here needs about 15MB total server-side with SSR enabled and shares type definitions across all layers (minus the database itself), so I can confirm neat things are at least technically feasible.
Hopefully I'll eventually get back to that part, maybe to come up with a nice way to do remote function calls without repeating boilerplate.

In any case, I'll see if I can give `lignin`, `lignin-dom`, `lignin-html`, and _maybe_ [`lignin-schema` (link to its _lib.rs_)](https://github.com/Tamschi/lignin-schema/blob/develop/src/lib.rs) a polish pass this week. I use that last one to validate element names and (basic) semantics so far, but would like to add per-element attributes and event names to it too.

There won't be any large changes to the APIs for a while I think, so feel free to start exploring, and please tell me how it goes when you do so. The upgrade to the a-bit-more-polished versions should be painless aside from being a "breaking" Semver increment and possibly bumping the required Rust version.

**Comment by platy** [comment]
I am more of a backend developer but since I have been one, more and more of the app has moved to the frontend and I've been following it there. I think great apps in the future will involve tight integration between frontend and backend components which don't treat them as separate things. I tried this with JS but also found either a lot of pain points or found that you had to stay between the lines to avoid problems (and end up spending as much time configuring webpack or similar as crafting the app).

I gave it a try earlier and managed to get it working on frontend and backend using your example above with a couple of tweaks. These were the issues I came up against and workarounds:

- the published lignin_schema depends on lignin v0.0.3 (so I used master)
- the trait `std::error::Error` is not implemented for `lignin_html::Error<'_>` and it's non-exhaustive which limits what can be done with it
- not yet implemented: Update attributes and event bindings. (I didn't need any for my test so I commented out the todo!)
- not yet implemented: Unhandled child node diff: (Multi(..)) (I only needed one node so I unwrapped it myself).

In terms of features, my first thoughts are (I need to look in to it more though):
- exporting a trait like `dodrio::Render` - it's not core vdom but could be very useful for interop
- possibly also a builder similar to `dodrio::builder` - in a separate crate
- ability to define a javascript serialisation of event handlers which can be output to html (I think I've seen this on another crate but I can't find it)

I'll try some more things out when I have time :)

**Comment by Tamschi** [comment]
> * the published lignin_schema depends on lignin v0.0.3 (so I used master)

Ah, I must have forgotten to publish that last time around.
It's a bit off, but I'll make sure to update it for the new release(s) here.

- - -

I ended up implementing slightly larger changes to this crate today, since I noticed the old event handler system was very prone to memory leaks.

I've made a pull request at https://github.com/Tamschi/lignin/pull/8 to discuss this more easily.
The remnants are a bit side-lined right now. They're still represented as (unreachable) `Node` variant, but I probably won't figure out what I need there before upgrading `lignin-dom` quite a bit.
